zhusuan/distributions/multivariate.py

```diff
@@ -29,6 +31,144 @@
 ]
 
 
+class MultivariateNormalTriL(Distribution):
+    """
+    The class of multivariate normal distribution, where covariance is
+    parameterized with the lower triangular matrix in Cholesky decomposition,
+
+        .. math :: L \\text{s.t.} LL^T = \Sigma.
+
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    :param mean: An N-D `float` Tensor. mean[..., i_{N-1}, :] corresponds to a
+        single distribution.
+    :param cov_tril: An (N+1)-D `float` Tensor. Each slice represents the lower
+        triangular matrix in the Cholesky decomposition.
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    :param is_reparameterized: A Bool. If True, gradients on samples from this
+        distribution are allowed to propagate into inputs, using the
+        reparametrization trick from (Kingma, 2013).
+    :param use_path_derivative: A bool. Whether when taking the gradients
+        of the log-probability to propagate them through the parameters
+        of the distribution (False meaning you do propagate them). This
+        is based on the paper "Sticking the Landing: Simple,
+        Lower-Variance Gradient Estimators for Variational Inference"
+    :param check_numerics: Bool. Whether to check numeric issues.
+    """
+
+    def __init__(self,
+                 mean,
+                 cov_tril,
+                 group_ndims=0,
+                 is_reparameterized=True,
+                 use_path_derivative=False,
+                 check_numerics=False,
+                 **kwargs):
+        self._check_numerics = check_numerics
+        self._mean, self._n_dim = assert_rank_at_least_one(
+            mean, 'MVNTriL.mean')
+        self._cov_tril, cov_tril_last_shapes = assert_rank_at_least(
+            cov_tril, 2, 'MVNTriL.cov_tril')
+        # Immediately evaluated if both args are constant
+        err_msg = 'MVNTriL.cov_tril should have compatible shapes with mean'
+        assert_ops = [tf.assert_equal(s, self._n_dim, [err_msg])
+                      for s in cov_tril_last_shapes]
+        assert_ops += [tf.assert_equal(
+            tf.shape(self._mean), tf.shape(self._cov_tril)[:-1], [err_msg])]
```

Comment:
Will this raise errors immediately when its static shape is not equal?

Reply:
If True, I suspect the error message may not be very informative?

Reply:
> Will this raise errors immediately when its static shape is not equal?
Yes for the first line. The second line doesn't raise when the shape isn't fully defined (e.g. `[None,5]` vs `[None,6]`). Will fix. 
Error messages can be improved with the third arg.